File: rules-interpreter/src/test/java/ReturnExprTest.java
import com.github.iamrenny.ruleflow.Workflow;
import com.github.iamrenny.ruleflow.vo.WorkflowResult;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Map;
import java.util.Set;

class ReturnExprTest {

    @Test
    public void givenRuleWhenReturningValueMustReturnX() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x + y * z  = 15 return expr(x)
                default allow
            end
        """;

        Workflow ruleEngine = new Workflow(workflow);
        WorkflowResult expectedResult = new WorkflowResult("test", "dummy", "item_a", "15");
        WorkflowResult result = ruleEngine.evaluate(Map.of(
            "x", 15,
            "y", 22,
            "z", 0
        ));

        Assertions.assertEquals(expectedResult, result);
    }

    @Test
    public void givenExpressionWithValidExpressionWhenReturningMustCalculateValue() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' bucket = '4' return expr(courier_model_response.score * city_features.percentile_q1)
                default 0
            end
        """;

        Workflow ruleEngine = new Workflow(workflow);
        WorkflowResult expectedResult = new WorkflowResult("test", "dummy", "item_a", "2.0", Set.of());
        WorkflowResult result = ruleEngine.evaluate(Map.of(
            "bucket", "4",
            "courier_model_response", Map.of("score", 0.5),
            "city_features", Map.of("percentile_q1", 4)
        ));

        Assertions.assertEquals(expectedResult, result);
    }

    @Test
    public void givenRuleWhenReturningExprMustReturnCorrectValue() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x + y * z  = 15 return expr(x + y)
                default allow
            end
        """;

        Workflow ruleEngine = new Workflow(workflow);
        WorkflowResult expectedResult = new WorkflowResult("test", "dummy", "item_a", "37.0");
        WorkflowResult result = ruleEngine.evaluate(Map.of(
            "x", 15,
            "y", 22,
            "z", 0
        ));

        Assertions.assertEquals(expectedResult, result);
    }
}

File: rules-interpreter/src/test/kotlin/ReturnExprTest.kt
import com.github.iamrenny.rulesflow.Workflow
import com.github.iamrenny.rulesflow.vo.WorkflowResult
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class ReturnExprTest {
    @Test
    fun `given a rule when returning a value must return x`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x + y * z  = 15 return expr(x)
                default allow
            end
        """

        val ruleEngine = Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "dummy", "item_a", "15"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to 15,
                    "y" to 22,
                    "z" to 0
                )
            )
        )
    }
    @Test
    fun `given an expression with a valid expression when returning must calculate value`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' bucket = '4' return expr(courier_model_response.score * city_features.percentile_q1)
                default 0
            end
        """
        val ruleEngine = Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "dummy", "item_a", "2.0000", warnings = setOf()),
            ruleEngine.evaluate(
                mapOf(
                    "bucket" to "4",
                    "courier_model_response" to mapOf("score" to 0.5),
                    "city_features" to mapOf("percentile_q1" to 4)
                )
            )
        )
    }


    @Test
    fun `given a rule when returning an expr must return correct value`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x + y * z  = 15 return expr(x + y)
                default allow
            end
        """

        val ruleEngine = Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "dummy", "item_a", "37.00"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to 15,
                    "y" to 22,
                    "z" to 0
                )
            )
        )
    }
}